From 8b203adae76d32066d10c5339bbde8025518234b Mon Sep 17 00:00:00 2001
From: Alexandre Jutras <alexandre.jutras@linaro.org>
Date: Tue, 12 Nov 2019 14:04:40 -0500
Subject: [PATCH 1/5] OpenCDMi: Retrieve sub-sample data from ocdm client

Signed-off-by: Alexandre Jutras <alexandre.jutras@linaro.org>
---
 OpenCDMi/FrameworkRPC.cpp | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/OpenCDMi/FrameworkRPC.cpp b/OpenCDMi/FrameworkRPC.cpp
index 7d82aca..8d07a12 100644
--- a/OpenCDMi/FrameworkRPC.cpp
+++ b/OpenCDMi/FrameworkRPC.cpp
@@ -269,12 +269,14 @@ namespace Plugin {
                             if (IsRunning() == true) {
                                 uint8_t keyIdLength = 0;
                                 const uint8_t* keyIdData = KeyId(keyIdLength);
+                                uint32_t SubSampleCount = SubSampleDataLength() / sizeof(uint32_t);
+                                uint32_t *SubSample = (uint32_t *)SubSampleData();
 
                                 int cr = _mediaKeys->Decrypt(
                                     _sessionKey,
                                     _sessionKeyLength,
-                                    nullptr, //subsamples
-                                    0, //number of subsamples
+                                    SubSample,
+                                    SubSampleCount,
                                     IVKey(),
                                     IVKeyLength(),
                                     Buffer(),
-- 
2.17.1

