From b43f67d78aff2ebb20fbfef384b8793cd569346c Mon Sep 17 00:00:00 2001
From: Alexandre Jutras <alexandre.jutras@linaro.org>
Date: Wed, 4 Mar 2020 20:21:45 -0500
Subject: [PATCH 5/5] OpenCDMi: Disable debug log in socket_server_helper

Signed-off-by: Alexandre Jutras <alexandre.jutras@linaro.org>
---
 OpenCDMi/helper/socket_server_helper.cpp | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/OpenCDMi/helper/socket_server_helper.cpp b/OpenCDMi/helper/socket_server_helper.cpp
index a2e9f00..ebbf696 100644
--- a/OpenCDMi/helper/socket_server_helper.cpp
+++ b/OpenCDMi/helper/socket_server_helper.cpp
@@ -94,7 +94,9 @@ int SocketServer::Connect(int f_SocketChannelId)
   socketAddress.sun_family = AF_UNIX;
   sprintf(&socketAddress.sun_path[1], "opencdm_fd_communication_channel_0x%08x", f_SocketChannelId);
 
+#ifdef ENABLE_DEBUG_LOG
   CDMI_DLOG() << "Binding socket (" << &socketAddress.sun_path[1] << ")" << std::endl;
+#endif
 
   if(bind(lSocketFd,  (struct sockaddr *)&socketAddress, sizeof(socketAddress)) < 0) {
     CDMI_ELOG() << "Failure to bind socket" << std::endl;
@@ -140,7 +142,9 @@ int SocketServer::Connect(int f_SocketChannelId)
     goto handle_error;
   }
 
+#ifdef ENABLE_DEBUG_LOG
   CDMI_DLOG() << "SocketServer::Connect(): Connection is established" << std::endl;
+#endif
 
 handle_error:
   if(status < 0) {
-- 
2.17.1

