From 0e0fa9b0d968205bf22b582d4f28765aa860bcbe Mon Sep 17 00:00:00 2001
From: Alexandre Jutras <alexandre.jutras@linaro.org>
Date: Fri, 20 Mar 2020 13:37:03 -0400
Subject: [PATCH 17/17] [OCDM] Update interface to IDRM.h

Signed-off-by: Alexandre Jutras <alexandre.jutras@linaro.org>
---
 MediaSession.h | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/MediaSession.h b/MediaSession.h
index 17167f3..1af2219 100644
--- a/MediaSession.h
+++ b/MediaSession.h
@@ -18,7 +18,7 @@
 #pragma once
 
 #include <cdm.h>
-#include <cdmi.h>
+#include "interfaces/IDRM.h"
 
 namespace CDMi
 {
@@ -74,7 +74,6 @@ public:
         int secureFd,
         uint32_t secureSize);
 
-
     virtual CDMi_RESULT ReleaseClearContent(
         const uint8_t *f_pbSessionKey,
         uint32_t f_cbSessionKey,
-- 
2.17.1

